nifpga/bitfile.py

```diff
@@ -75,27 +84,480 @@ def base_address_on_device(self):
+            data = bits_from_fpga & self._data_mask
+            return data
+
+        signed_bit_mask = 1 << (self._size_in_bits - 1)
+
+        def unpack_numeric_signed(bits_from_fpga):
+            data = bits_from_fpga & self._data_mask
+            if data & signed_bit_mask:
+                data = data ^ self._data_mask
+                data += 1
+                data *= -1
+            return data
+        self._unpack = unpack_numeric_signed if self._signed else unpack_numeric_unsigned
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datatype(self):
+        return self._datatype
+
+    @property
+    def size_in_bits(self):
+        return self._size_in_bits
+
+    @property
+    def is_c_api_type(self):
+        return True
+
+    def unpack_data(self, data):
+        return self._unpack(data)
+
+    def pack_data(self, data_to_pack, packed_data):
+        packed_data = packed_data << self._size_in_bits
+        return packed_data | (data_to_pack & self._data_mask)
+
+
+class _Float(object):
+    """ Handles packing and unpacking floating point values from the FPGA. """
+    def __init__(self, name, type_name):
+        self._name = name
+        if "SGL" == type_name:
+            self._size_in_bits = 32
+            self._datatype = DataType.Sgl
+        elif "DBL" == type_name:
+            self._size_in_bits = 64
+            self._datatype = DataType.Dbl
+        self._data_mask = (1 << self._size_in_bits) - 1
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datatype(self):
+        return self._datatype
+
+    @property
+    def size_in_bits(self):
+        return self._size_in_bits
+
+    @property
+    def is_c_api_type(self):
         return True
 
-    def _is_fifo_fxp(self, channel_xml):
-        datatype = channel_xml.find("DataType").find("SubType").text.title()
-        if str(DataType.Fxp).lower() not in datatype.lower():
-            return False
+    def unpack_data(self, data):
+        data = data & self._data_mask
+        if self._datatype == DataType.Sgl:
+            return ctypes.c_float.from_buffer(ctypes.c_uint(data)).value
+        if self._datatype == DataType.Dbl:
+            return ctypes.c_double.from_buffer(ctypes.c_ulonglong(data)).value
+
+    def pack_data(self, data_to_pack, packed_data):
+        if self._datatype == DataType.Sgl:
+            bits_to_pack = ctypes.c_uint.from_buffer(ctypes.c_float(data_to_pack)).value
+        if self._datatype == DataType.Dbl:
+            bits_to_pack = ctypes.c_ulonglong.from_buffer(ctypes.c_double(data_to_pack)).value
+        return (packed_data << self._size_in_bits) | bits_to_pack
+
+
+class _Bool(object):
+    """ Handles packing and unpacking bools. """
+    def __init__(self, name):
+        self._name = name
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datatype(self):
+        return DataType.Bool
+
+    @property
```

Comment:
I feel like both the size_in_bits and is_c_api_type should really exist under the DataType class, as it makes more sense there. It feels weird to have hardware specific information in our XML parser.

Reply:
Being an enum makes datatype difficult.  Its also public, so I hate to change it too much.  size_in_bits is complicated for cluster and FXP.  is_c_api_type could fit though.

_return_ctype is also getting more complicated.  For Cluster controls and indicators, the C type used is uint32_t, for FIFOs, it will be uint8_t.

I'm open to suggestions on this.  I ended up just creating extra type classes because changing DataType and considering compatibility gets difficult.